baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetValue.java

```diff
@@ -0,0 +1,99 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.CompileTimeConstantExpressionMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.LiteralTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "OptionalOrElseGetValue",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = SeverityLevel.WARNING,
+        summary = "If lambda passed to Optional#orElseGet returns a simple expression, use Optional#orElse instead")
+public final class OptionalOrElseGetValue extends BugChecker implements MethodInvocationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+    private static final Matcher<ExpressionTree> OR_ELSE_GET_METHOD =
+            MethodMatchers.instanceMethod().onExactClass("java.util.Optional").named("orElseGet");
+    private static final Matcher<ExpressionTree> COMPILE_TIME_CONSTANT = new CompileTimeConstantExpressionMatcher();
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        match:
+        {
+            if (!OR_ELSE_GET_METHOD.matches(tree, state)) {
+                break match;
+            }
+
+            ExpressionTree orElseGetArg = tree.getArguments().get(0);
+
+            if (orElseGetArg.getKind() != Tree.Kind.LAMBDA_EXPRESSION) {
+                break match;
+            }
+
+            LambdaExpressionTree lambdaExpressionTree = (LambdaExpressionTree) orElseGetArg;
+            LambdaExpressionTree.BodyKind bodyKind = lambdaExpressionTree.getBodyKind();
+
+            if (bodyKind != LambdaExpressionTree.BodyKind.EXPRESSION) {
+                break match;
+            }
+
+            ExpressionTree expressionBody = (ExpressionTree) lambdaExpressionTree.getBody();
+            if (COMPILE_TIME_CONSTANT.matches(expressionBody, state)
```

Comment:
Is the compile time constant check ever necessary? What does it catch that the following checks cannot?

Reply:
It covers expressions that may look more complicated than the other cases, but can be evaluated at compile time. For example `PREFIX + "_someSuffix"` or `MIN + SIZE * 1.5`.

Added a test for such case.